poly/joint.go

```diff
@@ -0,0 +1,286 @@
+	if err == nil {
+		r.Dealers = append(r.Dealers, dealer)
+	}
+	return resp, err
+}
+
+// ProduceSharedSecret will generate the sharedsecret relative to this receiver
+// it will throw an error if something is wrong such as not enough Dealers received
+func (r *Receiver) ProduceSharedSecret() (*SharedSecret, error) {
+	if len(r.Dealers) < 1 {
+		return nil, errors.New("Receiver has 0 Dealers in its data.Can't produce SharedSecret.")
+	}
+	pub := new(PubPoly)
+	//pub.InitNull(r.info.Suite, r.info.T, r.Dealers[0].Promise.PubPoly().GetB())
+	pub.InitNull(r.info.Suite, r.info.T, r.info.Suite.Point().Base())
+	share := r.info.Suite.Secret()
+	goodShare := 0
+	for index, _ := range r.Dealers {
+		// Only need T shares
+		if goodShare >= r.info.T {
+			break
+		}
+		// Compute secret shares of the shared secret = sum of the respectives shares of peer i
+		// For peer i , s = SUM fj(i)
+		// TODO WARNING : THIS IS TEST IMPLEMENTATION
+		// In reality we should receive a NEW state struct from the dealer which is Certified so
+		// we can call RevealShare
+		// In testing we don't care about malicous yet so we just create one here
+		state := new(State).Init(*r.Dealers[index].Promise)
+		s, e := state.RevealShare(r.index, r.Key)
+		//s, e := r.Dealers[index].State.RevealShare(r.index, r.Key)
+		if e != nil {
+			//TODO error handling function not implemented right now. Only used for testing / comparison.
+			// We must be able to tell which share failed and to implement the broadcast of that error to others receiver
+			// so they reconstruct the private polynomial of the malicious dealer and set their share themself
+			return nil, errors.New(fmt.Sprintf("Receiver %d could not reveal its share from Dealer %d promise : %v", r.index, index, e))
+		}
+		share.Add(share, s)
+
+		// Compute shared public polynomial = SUM of indiviual public polynomials
+		pub.Add(pub, r.Dealers[index].Promise.PubPoly())
+
+		goodShare += 1
+	}
+
+	if goodShare < r.info.T {
+		return nil, errors.New("Not enough shares received by the Receiver to construct its own share of the shared secret")
+	}
+
+	if val := pub.Check(r.index, share); val == false {
+		return nil, errors.New("Receiver's secret share of the shared secret could not be checked against the shared polynomial")
+	}
+
+	return &SharedSecret{
+		Pub:   pub,
+		Share: &share,
+		Index: r.index,
+	}, nil
+}
+
+// MARSHALLING side
+
+// PolyInfo marshalling :
+// It is important to note that it does not Marshal / Unmarshal
+// the SUITE used. The suite is generated by default in constants.go
+// If you want to change to another suite, just change SUITE = ....
+func (p *PolyInfo) MarshalSize() int {
+	return util.IntSize(p.N) +
+		util.IntSize(p.R) +
+		util.IntSize(p.T)
+}
+
+func (p *PolyInfo) Equal(p2 PolyInfo) bool {
+	return p.N == p2.N && p.R == p2.R && p.T == p2.T && p.Suite.String() == p2.Suite.String()
+}
+
+func (p *PolyInfo) MarshalBinary() ([]byte, error) {
```

Comment:
And if you do for some reason want/need to implement your own MarshalBinary function here, it would be cleaner and more efficient to do it something like:

var buf bytes.Buffer
var val [3]uint64 = {(uint64)p.T, (uint64)p.R, (uint64)p.N}
err := abstract.Write(&buf, &val, suite)
...